pkg/operator/watchdog.go

```diff
@@ -0,0 +1,172 @@
+package operator
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/ratelimit"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/dapr/dapr/utils"
+)
+
+const (
+	sidecarContainerName     = "daprd"
+	daprEnabledAnnotationKey = "dapr.io/enabled"
+)
+
+// DaprWatchdog is a controller that periodically polls all pods and ensures that they are in the correct state.
+// This controller only runs on the cluster's leader.
+// Currently, this ensures that the sidecar is injected in each pod, otherwise it kills the pod so it can be restarted.
+type DaprWatchdog struct {
+	enabled           bool
+	interval          time.Duration
+	maxRestartsPerMin int
+
+	client         client.Client
+	restartLimiter ratelimit.Limiter
+}
+
+// NeedLeaderElection makes it so the controller runs on the leader node only.
+// Implements sigs.k8s.io/controller-runtime/pkg/manager.LeaderElectionRunnable .
+func (dw *DaprWatchdog) NeedLeaderElection() bool {
+	return true
```

Comment:
Why even do this?

Reply:
Operator service can run in HA mode where there are 3 instances. We need to make sure that only one instance of operator is killing pods, otherwise there are race conditions and pods can be killed too many times (and in general there can be too many invocations of the K8s APIs).

Luckily for us, the operator service already implemented a Controller Manager (see https://github.com/dapr/dapr/pull/4764/files#diff-99c0290799daf9abc6240df64063e20bfaf67b371577b67ac7eec6f4725622ffR91 ) , so I plugged this watchdog as a controller. 

All controllers need to implement the `sigs.k8s.io/controller-runtime/pkg/manager.Runnable` interface, which includes the `Start` method below.

This `NeedLeaderElection` method is an optional one to implement `sigs.k8s.io/controller-runtime/pkg/manager.LeaderElectionRunnable` which makes it so the controller only runs on the instance of the service that K8s elects as leader.

So if there are 3 instances of Operator, the watchdog runs only on one. If the leader dies, another instance will re-start the operator. I have manually verified this behavior.

PS: if the watchdog was configured to run only `once` and the leader of the operator service dies, when a new leader is elected the watchdog be triggered again - I don't think this should be an issue however.

Reply:
OK. I understand now. Thanks.